### ISSUE TITLE
RFC: Generic generic parameters

### ISSUE THREAD

**Issue by bpr** [nim-lang/Nim#3856]
Nim doesn't allow parameters of generic to be generic themselves. The lack of this feature has come up a few times, sometimes in the context of implementing something from functional programming like a
flatMap or monadic API, and also as with C++ in parameterizing containers. The usual workaround in the
absence of this feature is to pass in only instantiated types to generics.

This feature exists in C++, D, and Scala, and it is one of the highly desired missing features from Rust, see 'higher kinded types', or HKT, where it's perhaps more necessary than in Nim since Rust uses algebraic data types for errors rather than exceptions. Having this, variadic generics, and working static[T], and concepts will close the gap between Nim and C++ 17.

**Comment by vegansk** [comment]
:+1: for HKT!

**Comment by Araq** [comment]
Note that higher kinded types are not as required since you can simply underspecify generic constraints in Nim (Nim's generics are close to C++ templates in this regard).

**Comment by bpr** [comment]
The feature, while not required, is consistent with (my view of) the Nim way, works especially well with concepts, and increases the expresiveness of the type system. That you can get around their omission by specifying less strikes me as making a virtue of a necessity. Let's take the simplest example in pseudo-Nim

```
type
  Array[T] = concept a
    a.len is Ordinal
    get(a, Ordinal) is T
    for value in a:
      type(value) is T

  Map[K, V, A: Array[T]] = object
    keys: A[K]
    values: A[V]
```

The meaning is clear, and the variations on this (allowing another generic array param to allow different arrays to back keys and values) are straightforward.

**Comment by Xe** [comment]
Isn't this already [the case](http://nim-lang.org/docs/manual.html#generics-concepts)?

**Comment by vegansk** [comment]
@Xe, no. The previous example is written in ``pseudo-Nim``. You can't implement ``Map`` like this in real Nim. How I wish to be wrong, but I can't find any solution.

**Comment by mratsim** [comment]
I would love that.

I have a base method declared with a generic type TT (TT is expected to be Container[T])

But in some implementations I need to get access to the T subtype of TT for proc like `newSeq[T](foo)` to return a new container.

**Comment by hlaaftana** [comment]
This works:

```nim
type
  Array[T] = concept a
    a.len is Ordinal
    #get(a, Ordinal) is T
    for value in a:
      value is T

  Map[K, V] = concept m
    type A = Array
    m.keys is A[K]
    m.values is A[V]

template mapImpl(A: typedesc): auto =
  type MapImpl[Key, Value] = object
    keys: A[Key]
    values: A[Value]
  
  assert MapImpl is Map
  
  MapImpl

proc get[T](arr: openarray[T], idx: Ordinal): T = arr[idx]

type
  SeqMap = mapImpl(seq)

iterator pairs(m: Map): (m.K, m.V) =
  let ks = iterator(keys: type(m.keys)): m.K =
    for k in keys:
      yield k
  let vs = iterator(values: type(m.values)): m.V =
    for v in values:
      yield v
  while true:
    let p = (ks(m.keys), vs(m.values))
    if ks.finished or vs.finished: break
    yield p

proc toSeqMap[K, V](ps: openarray[(K, V)]): SeqMap[K, V] =
  result.keys = newSeq[K](ps.len)
  result.values = newSeq[V](ps.len)
  for i, p in ps.pairs:
    (result.keys[i], result.values[i]) = p

var sm = {"a": 1, "b": 2}.toSeqMap

for k, v in sm:
  echo "key: ", k, ", value: ", v
```

Comment out the `get(a, Ordinal) is T` line and it breaks. Change it with `a[Ordinal] is T` and it breaks again because the system module says `T[I] is T` for whatever reason
